### PR TITLE
Merge memindex

### PR DESCRIPTION
**Pull request by thani-sh**

### CODE REVIEW

**Comment by thani-sh** [comment]
### TODO
- [ ] `TestFindElementsInMiddleRoot` fails sometimes, kinda random
- [x] Have to fix persisting the tree to file and getting it back

**Comment by thani-sh** [comment]
Persisting the whole index at once and getting it back works for now.

**Comment by arunoda** [comment]
Okay. That's great.
Cool.

File: memindex.go [hunk 1]
```diff
@@ -1,26 +1,24 @@
...
 }
 
 // Base struct of the MemIndex
 // It contains an element called `root` which is the starting point of the tree
-// orderedKeys contains all the keys in this index in ordered fashion
+// keys contains all the keys in this index in ordered fashion
 type MemIndex struct {
-	root        *IndexElement
-	orderedKeys []string
+	Root *IndexElement
```
Comment (arunoda) [inline memindex.go:20]: I make these simple because they are private. 
Isn't it a patter?

Comment (thani-sh) [inline memindex.go:20]: json/gob encoders ignore private fields

Comment (arunoda) [inline memindex.go:20]: Ah Owen. Yes yes.
Btw: that test failures could be fixed by sorting positions array.

**Comment by thani-sh** [comment]
Merged

**Pull request closed without merging**